Add config parameter to change per-container stop timeout during daemon shutdown

yongtang opened the pull request:
**\- What I did**

This fix tries to add a flag `--stop-timeout` to specify the timeout value (in seconds) for the container to stop before SIGKILL is issued. If stop timeout is not specified then the default timeout (10s) is used.

**\- How I did it**

The `--stop-timeout` has been added to `docker create` and `docker run`.

**\- How to verify it**

Additional test cases have been added to cover the change.

**\- Description for the changelog**

Added a lag `--stop-timeout` to specify the timeout value (in seconds) for the container to stop.

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix is related to #22471.

**NOTE: Pull requests have been created in engine-api:**
https://github.com/docker/engine-api/pull/256
https://github.com/docker/engine-api/pull/271

**NOTE: Another pull request (#23036) has been opened to add `--shutdown-timeout` to daemon**

Signed-off-by: Yong Tang yong.tang.github@outlook.com

pataquets commented:
Will this option match the corresponding ```docker stop --time=-1``` parameter value to enable waiting undefinitely for the container to stop?

yongtang commented:
@pataquets Yes if the option is passed with `-1` then the daemon will wit indefinitely for clean stop of the containers.

thaJeztah commented:
I'm wondering if this should be a daemon level option, or something that should be specified per container (thinking of the `--stop-signal` option we added on `docker run`, and in the Dockerfile).

I like the simplicity of this PR, just thinking out loud if the timeout should be applied to all containers, or only to specific ones.

LK4D4 commented:
I think restarts without container stops(in 1.12) will deprecate this.

thaJeztah commented:
@LK4D4 that was mentioned in the issue, but it likely won't help if the service has to be stopped, or a server rebooted

yongtang commented:
@thaJeztah @LK4D4 Thanks for the review.

In the existing docker code, the timeout of container stop before SIGKILL during the shutdown is controlled by two places:
1. shutdownContainer() (with a timeout of 10s):
https://github.com/docker/docker/blob/e16753ce192cc80d3e207d7b3063a9dab36983cb/daemon/daemon.go#L860-L864
2. shutdownDaemon() (with a timeout of 15s = 10s + 5s):
https://github.com/docker/docker/blob/e16753ce192cc80d3e207d7b3063a9dab36983cb/cmd/dockerd/daemon.go#L290-L294

When a shutdown of daemon is issued, the daemon will:
1. Send SIGTERM to each container, wait for 10s, and send SIGKILL if containers hasn't been stopped yet.
2. At the same time, wait for all containers to stop or 15s (max), then exit.

There are potentially two variables. One is the timeout (10s) for each container to stop (with SIGTERM) before send SIGKILL. The other is the timeout (15s) for daemon to exit.

In this PR, there is only one variable `--shutdown-timeout`. Basically each container will wait for `shutdown-timeout` before SIGKILL is issued. At the same time, the daemon will wait for `shutdown-time + 5s` for all containers to stop, before exit. The added value of `5s` is hardcoded for the moment in this PR.

I think the `per-container-stop-timeout` makes sense. Maybe we could set a `per-container-stop-timeout` for each container, then add a `shutdown-timeout` for the daemon to wait overall?

Any suggestions would be appreciated.

icecrime commented:
I think this makes sense (regardless of the potential ability for 1.12 to restart the daemon without restarting containers).

*Every such PR makes me think we should really stop adding command line flags and support a configuration file only...*

yongtang commented:
@thaJeztah @LK4D4 @icecrime To implement a `per-container-stop-timeout` the following items are likely to be needed:
* Add a field `StopTimeout` to engine-api/types/container/config.go (`StopSignal` already exists)
* Add a flag of `--stop-timeout=10` to `docker run` and `docker create`
* Add a flag of `--shutdown-timeout=15` to `docker daemon` for the wait time before forcefully exit daemon.
* When the daemon is shutdown, each container will be stopped based on the `StopTimeout` in config before issuing a SIGKILL. In addition, daemon will wait for `--shutdown-timeout` before forcefully exit.

* The only potential uncertainly is the `docker stop --time=10`. When `docker stop --time=10` is executed, the stop timeout will be the value passed in command which already has a default value. I think it makes sense for `docker stop` to respect `StopTimeout` if the `--time` flag is not given.

I will move forward to implement the above items and update the PR. In the meantime, please let me know if you have any suggestions or feedbacks.

thaJeztah commented:
Perhaps we can combine `stop-signal` with `stop-timeout`, similar to the `--restart=always:10`, so something like `docker run --stop-signal=kill:30` (downside is that you'd always have to specify the *signal* as well)

yongtang commented:
@thaJeztah @LK4D4 @icecrime I updated the PR and now `--stop-signal` could take both a signal and a timeout e.g., `--stop-signal=SIGTERM:10`.

The `StopTimeout` in Config is defined as `*int`. This is to allow the case where value `StopTimeout` is not assigned (then default 10s will be used).

Note: we cannot use `-1` or `0` as the default value of StopTimeout because `-1` means infinity and `0` means no waiting.

One thing I am not so sure is about how to deal with `docker stop`. Since `docker stop -t` always has a default value, it seems like the `StopTimeout` is useless for `docker stop`.

If we want to respect the `StopTimeout` even for `docker stop`, I think we could change the behavior so that:
* If `docker stop` does not have the flag `-t` then we will use `StopTimeout`
* Otherwise we will use the value provided by `docker stop -t`

Any suggestion suggestions or feedbacks on how to deal with `docker stop`?

mattyb commented:
It may be nice to combine this with --start-timeout, per docker/docker#22226

jhowardmsft commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -2332,3 +2332,51 @@ func (s *DockerDaemonSuite) TestBuildOnDisabledBridgeNetworkDaemon(c *check.C) {
+
+// Test case for #22471
+func (s *DockerDaemonSuite) TestDaemonShutdownTimeout(c *check.C) {
+	testRequires(c, SameHostDaemon, DaemonIsLinux)
```
Why Linux only?

yongtang replied:
@jhowardmsft Thanks for point that out. I will update the PR to cover Windows as well.

thaJeztah commented:
We just discussed this; and there's more people that like having a separate flag for `--stop-timeout` after all (sorry).

Also we'd like to have the per-container feature first, and a separate PR for the default (daemon) option.
Can you do that? Thanks!

yongtang committed:
Add config parameter to change per-container stop timeout during daemon shutdown

This fix tries to add a flag `--stop-timeout` to specify the timeout value
(in seconds) for the container to stop before SIGKILL is issued. If stop timeout
is not specified then the default timeout (10s) is used.

Additional test cases have been added to cover the change.

This fix is related to #22471. Another pull request will add `--shutdown-timeout`
to daemon for #22471.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @thaJeztah @jhowardmsft I split the pull request into two. This pull request will cover the 'per-container' stop timeout case and the pull request #23036 will cover daemon timeout case. Please let me know if there are any issues.

thaJeztah commented:
Thanks @yongtang! I'm moving this to code review

yongtang commented:
Thanks all for the code review and suggestions. A pull request for engine-api has been created:
https://github.com/docker/engine-api/pull/256
Please let me know if there are any issues.

thaJeztah commented:
Thanks @yongtang, https://github.com/docker/docker/issues/23227 was just merged, so you can rebase to fix the `vendor` CI 👍

yongtang commented:
Thanks @thaJeztah. The pull request has been rebased and updated. Please let me know if there are any issues.

cpuguy83 commented on pkg/signal/signal.go:
```diff
@@ -11,6 +11,11 @@ import (
 
+const (
+	// DefaultStopTimeout is the timeout (in seconds) for the syscall signal used to stop a container.
+	DefaultStopTimeout = 10
```
nit, this doesn't seem like the correct place.
Probably should go in either the daemon package or the container package.

yongtang replied:
Thanks @cpuguy83. I tried to move the `DefaultStopTime` to container package. However, it caused a circular import issue in `runconfig/opts/parse.go` as `runconfig` is used by `container` package. Let me investigate further to see if there is a way to get around this issue.

cpuguy83 replied:
I see. Probably best to leave the default off the CLI argument.
*Might* need to make `StopTimeout` a pointer so we know the difference between 0 and not set... but maybe 0 doesn't make sense here anyway.

thaJeztah replied:
Ah, yes, if possible, I prefer the CLI to not have to send anything if it's the default

yongtang commented:
Thanks @thaJeztah @cpuguy83.

I did some investigation. It turns out that, if I use the following (with value `0`):
```
flStopTimeout       = cmd.Int([]string{"-stop-timeout"}, 0, "Timeout (in seconds) to stop a container")
```
then `docker run --help` will return:
```
...
  --stop-timeout                  Timeout (in seconds) to stop a container
...
```

Also, if user does not specify `--stop-timeout=xxx` in `docker run`, then the following line:
```
	if cmd.IsSet("-stop-timeout") {
 		config.StopTimeout = flStopTimeout
 	} 
```
will correctly assign `config.StopTimeout` to `null`, which will be correctly picked up and assigned with the default time out value (10s) by the daemon.

Based on the observation, I changed the value to `0` in `runconfig` and moved `DefaultStopTimeout = 10` to `container` package.

The pull request has been updated with the above mentioned changes. Please let me know if there are any other issues.

thaJeztah commented:
Thanks @yongtang!

Thanks! Testing this, I noticed that we're not there yet (at least, I think
additional changes have to be made).

When doing `docker stop <containerId>`, the default 10 seconds is always applied,
because the client sends it;

```bash
docker run --stop-timeout=5 -dit --name five busybox
a28d9bc1202ac77e899122e9331b125c0331a49e5cdd5f32d3a1be505a90756a

docker run --stop-timeout=15 -dit --name fifteen busybox
aeb99e32e1103624bf5e28a06abfc52b644baefe8fa9c47d9794680e199c3e63

docker run -dit --name defaulty busybox
8066899f06eb273e77cf1f7c06244d3b85aa7415cc6cc15bf48e1e85f3a6823d

docker stop five fifteen defaulty
```
```
DEBU[0178] Calling POST /v1.24/containers/five/stop?t=10
DEBU[0178] Sending 15 to a28d9bc1202ac77e899122e9331b125c0331a49e5cdd5f32d3a1be505a90756a
INFO[0188] Container a28d9bc1202ac77e899122e9331b125c0331a49e5cdd5f32d3a1be505a90756a failed to exit within 10 seconds of signal 15 - using the force

DEBU[0189] Calling POST /v1.24/containers/fifteen/stop?t=10
DEBU[0189] Sending 15 to aeb99e32e1103624bf5e28a06abfc52b644baefe8fa9c47d9794680e199c3e63
INFO[0199] Container aeb99e32e1103624bf5e28a06abfc52b644baefe8fa9c47d9794680e199c3e63 failed to exit within 10 seconds of signal 15 - using the force

DEBU[0199] Calling POST /v1.24/containers/defaulty/stop?t=10
DEBU[0199] Sending 15 to 8066899f06eb273e77cf1f7c06244d3b85aa7415cc6cc15bf48e1e85f3a6823d
INFO[0209] Container 8066899f06eb273e77cf1f7c06244d3b85aa7415cc6cc15bf48e1e85f3a6823d failed to exit within 10 seconds of signal 15 - using the force
```

Doing a shutdown of the daemon correctly applies the timeouts set;

```
INFO[3797] Container a28d9bc1202ac77e899122e9331b125c0331a49e5cdd5f32d3a1be505a90756a failed to exit within 5 seconds of signal 15 - using the force
INFO[3802] Container 8066899f06eb273e77cf1f7c06244d3b85aa7415cc6cc15bf48e1e85f3a6823d failed to exit within 10 seconds of signal 15 - using the force
INFO[3807] Container aeb99e32e1103624bf5e28a06abfc52b644baefe8fa9c47d9794680e199c3e63 failed to exit within 15 seconds of signal 15 - using the force
```


I think we should have `docker stop` not send a value if it's the default,
and let the daemon apply either the default value, or the value that was
set on the container (through `--stop-timeout`), wdyt?

yongtang commented:
Thanks @thaJeztah. I think it is possible to achieve the behavior. It may involve updating engine-api. Let me give it a try.

yongtang committed:
Update `docker stop` and `docker restart` to allow not specifying timeout and use the one specified at container creation time.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@thaJeztah I updated the pull request to take into consideration the case where `docker stop` does not specify the timeout. In that case, a null is passed and, eventually reach the daemon so that daemon can use the config value specified at the container creation time.

Please take a look and I will create a separate pull request to engine-api if the update fits the behavior.

thaJeztah commented:
Thanks @yongtang, that's exactly what I had in mind; it picks the value set by the container, unless I override it, and explicitly set the time;

Just "stop";
```
docker stop fifteen
DEBU[0354] Calling POST /v1.24/containers/fifteen/stop
DEBU[0354] Sending 15 to 105b3e0b5e0bd213d952042f95671f66cf0a254318baaf24c5b4ab6b88e3190d
INFO[0369] Container 105b3e0b5e0bd213d952042f95671f66cf0a254318baaf24c5b4ab6b88e3190d failed to exit within 15 seconds of signal 15 - using the force
```

Override the timeout:

```
docker stop --time=5 fifteen
DEBU[0417] Calling POST /v1.24/containers/fifteen/stop?t=5
DEBU[0417] Sending 15 to 105b3e0b5e0bd213d952042f95671f66cf0a254318baaf24c5b4ab6b88e3190d
INFO[0422] Container 105b3e0b5e0bd213d952042f95671f66cf0a254318baaf24c5b4ab6b88e3190d failed to exit within 5 seconds of signal 15 - using the force
```

thaJeztah commented:
However, I noticed a fun thing; if the timeout of the container is longer than the default timeout (probably the timeout of the daemon), then the container is never killed:

Start a container with a 15 seconds timeout:
```
docker run --stop-timeout=15 -dit --name fifteen busybox
```

(logs)
```
DEBU[0077] Calling POST /v1.24/containers/create?name=fifteen
DEBU[0077] form data: {"AttachStderr":false,"AttachStdin":false,"AttachStdout":false,"Cmd":null,"Domainname":"","Entrypoint":null,"Env":[],"HostConfig":{"AutoRemove":false,"Binds":null,"BlkioDeviceReadBps":null,"BlkioDeviceReadIOps":null,"BlkioDeviceWriteBps":null,"BlkioDeviceWriteIOps":null,"BlkioWeight":0,"BlkioWeightDevice":null,"CapAdd":null,"CapDrop":null,"Cgroup":"","CgroupParent":"","ConsoleSize":[0,0],"ContainerIDFile":"","CpuCount":0,"CpuPercent":0,"CpuPeriod":0,"CpuQuota":0,"CpuShares":0,"CpusetCpus":"","CpusetMems":"","Devices":[],"DiskQuota":0,"Dns":[],"DnsOptions":[],"DnsSearch":[],"ExtraHosts":null,"GroupAdd":null,"IOMaximumBandwidth":0,"IOMaximumIOps":0,"IpcMode":"","Isolation":"","KernelMemory":0,"Links":null,"LogConfig":{"Config":{},"Type":""},"Memory":0,"MemoryReservation":0,"MemorySwap":0,"MemorySwappiness":-1,"NetworkMode":"default","OomKillDisable":false,"OomScoreAdj":0,"PidMode":"","PidsLimit":0,"PortBindings":{},"Privileged":false,"PublishAllPorts":false,"ReadonlyRootfs":false,"RestartPolicy":{"MaximumRetryCount":0,"Name":"no"},"SecurityOpt":null,"ShmSize":0,"StorageOpt":{},"UTSMode":"","Ulimits":null,"UsernsMode":"","VolumeDriver":"","VolumesFrom":null},"Hostname":"","Image":"busybox","Labels":{},"NetworkingConfig":{"EndpointsConfig":{}},"OnBuild":null,"OpenStdin":true,"StdinOnce":false,"StopTimeout":15,"Tty":true,"User":"","Volumes":{},"WorkingDir":""}
DEBU[0077] container mounted via layerStore: /var/lib/docker/aufs/mnt/4cc2e835903d71e58e06229e3e6ddc677bb80710edcfe0f7a7e5f0a2966b7078
DEBU[0078] Calling POST /v1.24/containers/bd5b3c158853c458c3d0b0f59845f70dc36d78df643a3324c8ee866625c52035/start
DEBU[0078] container mounted via layerStore: /var/lib/docker/aufs/mnt/4cc2e835903d71e58e06229e3e6ddc677bb80710edcfe0f7a7e5f0a2966b7078
DEBU[0078] Assigning addresses for endpoint fifteen's interface on network bridge
DEBU[0078] RequestAddress(LocalDefault/172.18.0.0/16, <nil>, map[])
DEBU[0078] Assigning addresses for endpoint fifteen's interface on network bridge
DEBU[0078] Programming external connectivity on endpoint fifteen (6f6636fd63431aeeadbb894c7bfb2f0de6f6a766f563ea786f33d02a3da4e49a)
DEBU[0078] sandbox set key processing took 104.841449ms for container bd5b3c158853c458c3d0b0f59845f70dc36d78df643a3324c8ee866625c52035
DEBU[0078] received containerd event: &types.Event{Type:"start-container", Id:"bd5b3c158853c458c3d0b0f59845f70dc36d78df643a3324c8ee866625c52035", Status:0x0, Pid:"", Timestamp:0x57569fb4}
DEBU[0078] event unhandled: type:"start-container" id:"bd5b3c158853c458c3d0b0f59845f70dc36d78df643a3324c8ee866625c52035" timestamp:1465294772
```

Stop the daemon (^C):

```
^CINFO[0128] Processing signal 'interrupt'
DEBU[0128] starting clean shutdown of all containers...
DEBU[0128] stopping bd5b3c158853c458c3d0b0f59845f70dc36d78df643a3324c8ee866625c52035
DEBU[0128] Sending 15 to bd5b3c158853c458c3d0b0f59845f70dc36d78df643a3324c8ee866625c52035
INFO[0143] Container bd5b3c158853c458c3d0b0f59845f70dc36d78df643a3324c8ee866625c52035 failed to exit within 15 seconds of signal 15 - using the force
DEBU[0143] Sending 9 to bd5b3c158853c458c3d0b0f59845f70dc36d78df643a3324c8ee866625c52035
ERRO[0143] Force shutdown daemon
DEBU[0143] containerd connection state change: SHUTDOWN
INFO[0143] stopping containerd after receiving terminated
```

It looks like the daemon either doesn't send the timeout, or doesn't wait for sending it before shutting down itself. The container is still running after the daemon is stopped;

```
ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.1  20256  3192 pts/0    Ss   09:48   0:00 bash
root       946  0.0  0.1 259700  2828 ?        Sl   10:19   0:00 docker-containerd-shim bd5b3c158853c458c3d0b0f59845f70dc36d78df643a3324c8ee866625c52035 /var/run/docker/libcontainerd/bd5b3c158853c458c3d0b
root       962  0.0  0.0   1208     4 pts/0    Ss+  10:19   0:00 sh
root       974  0.0  0.1  17500  2076 pts/0    R+   10:21   0:00 ps aux
```

And re-appears after the daemon is started again;

```
docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
bd5b3c158853        busybox             "sh"                2 minutes ago       Up 2 minutes                            fifteen
```

thaJeztah commented:
So, indeed; the daemon doesn't wait for the timeout set by the containers,
and if their timeout is longer than 15 seconds (daemon default), the daemon
shuts down without killing the containers:

```
$ docker run --stop-timeout=20 -dit --name twenty busybox
146d818d38239b1164d0021f2bc8cd64d58112540d34ac3ca8f283ec7ec46b48

docker run --stop-timeout=25 -dit --name twentyfive  busybox
476494bbf2c0d36274ded855c097bf6bc7e58fe95264e65ac48f066987a3d970
```

```
^CINFO[0037] Processing signal 'interrupt'
DEBU[0037] starting clean shutdown of all containers...
DEBU[0037] stopping 476494bbf2c0d36274ded855c097bf6bc7e58fe95264e65ac48f066987a3d970
DEBU[0037] Sending 15 to 476494bbf2c0d36274ded855c097bf6bc7e58fe95264e65ac48f066987a3d970
DEBU[0037] stopping 146d818d38239b1164d0021f2bc8cd64d58112540d34ac3ca8f283ec7ec46b48
DEBU[0037] Sending 15 to 146d818d38239b1164d0021f2bc8cd64d58112540d34ac3ca8f283ec7ec46b48
ERRO[0052] Force shutdown daemon
DEBU[0052] containerd connection state change: SHUTDOWN
INFO[0052] stopping containerd after receiving terminated
```

thaJeztah commented:
@yongtang oh! just realized you mentioned the "default timeout" on `docker stop` in an earlier comment (https://github.com/docker/docker/pull/22566#issuecomment-221460299), I completely overlooked that earlier

yongtang commented:
Thanks @thaJeztah. The shutdown timeout is a combination of container and daemon. When daemon tries to shutdown, it will wait for `15s` before exit. However, this `15s` **implicitly** hints that the container stop timeout is `10s`.

The current implementation of this PR didn't automatically use this implicit value of `10s`. Instead, the stored stopTimeout is used.

I will update the PR shortly to use this implicit `10s` for daemon shutdown so the behavior is OK after this PR.

Though the `10s` fix might be temporary as this this PR is related to #23036.

#23036 will change the shutdown timeout of daemon (instead of `15s`) and that will impact the implicit `10s`.

thaJeztah commented:
Yes, I assumed that was the case; we should consider;

- the daemon timeout to be set to a longer time if a container has a longer timeout than the default
- not sure; what to do if a container has a timeout `-1` (indefinitely)? should this always block the daemon from stopping?
- if we *don't* raise the timeout, the container should probably still be killed?

Also, I noticed that `docker restart <container>` has the same issue as `docker stop`; it also passes the default timeout from the client, and ignores the timeout set on the container.

yongtang commented:
Thanks @thaJeztah I updated the `docker restart <container>` so it now behaves similarly with `docker stop <container>`.

I also capped the timeout to 10s when stopping containers at daemon shutdown. It ma need to be changed when #23036 is in play.

For daemon shutdown, currently docker is doing:
1. Issue shutdownContainer to stop all containers. If container failed to stop in 10s, then kill it.
2. Separately wait for 15s then exit directly.

If we want to make sure all containers stop when daemon shutdown, then we will need to pass a timeout at the shutdownContainer. (Daemon;s 15s shutdown timeout is done separately).

I think we can do two ways:
1. Scan through all containers, and find the longest timeout value say, `20s`, and let daemon wait for `20s + 5s` before exit.
2. Assume daemon timeout is `20s`, then cap all container stop timeout value to be max of `20s - 5s`.

We could do either way and the same could be applied to `-1` case.

I personally would see daemon level shutdown timeout as the `hard timeout` ( see 2) could be easier to deal with. Though I can do the implemention for either ways.

thaJeztah commented:
Thanks again @yongtang, and sorry for all the discussion here; I'm *inclined* to say that we should scan through all containers (or, since we're already going through all containers, if the container timeout is longer than the current timeout, raise it), after all, ignoring the custom timeout would make this feature not really useful.

However, I don't know what others think here; so I'll try to get some maintainers to look at it ASAP so that we can move forward.

yongtang commented:
@thaJeztah I rebased the pull request to fix the conflict. The implementation has also been updated. Right now, the daemon shutdown timeout is taken to scan through all containers and use the "max" StopTimeout from containers.

Please take a look and let me know if there are any issues, or we would like to have a different way of handling the shutdown timeout.

thaJeztah commented:
Looks like something is failing, not sure why;

```
22:23:34 # github.com/docker/libnetwork/osl
22:23:34 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:21: undefined: IfaceOption
22:23:34 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:42: undefined: NeighOption
22:23:34 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:67: undefined: NeighOption
22:23:34 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:71: undefined: NeighOption
22:23:34 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:77: undefined: IfaceOption
22:23:34 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:80: undefined: IfaceOption
22:23:34 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:83: undefined: IfaceOption
22:23:34 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:86: undefined: IfaceOption
22:23:34 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:91: undefined: IfaceOption
22:23:34 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:94: undefined: IfaceOption
22:23:34 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:94: too many errors
```

thaJeztah commented:
I tested this locally, and it works as I would expect, i.e., 

- both `docker restart` and `docker stop` use the timeout set by the container
- stopping the daemon also takes that time into account
- using `--time=something` on `docker stop` or `docker restart` overrides the timeout that's set by the container

ping @cpuguy83 @vdemeester other than the failing CI, does this behavior LGTY?

vdemeester commented:
I think it makes sence yes :)

yongtang commented:
Thanks @thaJeztah @vdemeester @cpuguy83. I created a pull request in engine-api to cover the needed changed in engine-api:
https://github.com/docker/engine-api/pull/271

Please let me know if there are any issues.

yongtang commented:
The janky failure seems to be timeout. Though experimental looks like for real. Not sure why the following linking error appears but will do some investigation.
```
04:00:15 # github.com/docker/libnetwork/osl
04:00:15 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:21: undefined: IfaceOption
04:00:15 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:42: undefined: NeighOption
04:00:15 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:63: undefined: IfaceOption
04:00:15 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:70: undefined: NeighOption
04:00:15 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:74: undefined: NeighOption
04:00:15 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:80: undefined: IfaceOption
04:00:15 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:83: undefined: IfaceOption
04:00:15 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:86: undefined: IfaceOption
04:00:15 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:89: undefined: IfaceOption
04:00:15 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:92: undefined: IfaceOption
04:00:15 vendor/src/github.com/docker/libnetwork/osl/sandbox.go:92: too many errors
04:00:24 Build step 'Execute shell' marked build as failure
```

yongtang commented:
The investigation is ongoing. It seems libnetwork does not really handle darwin yet experimental in docker somehow tries to build on darwin with libnetwork, thus caused the failure.

Will spend more time on this issue.

thaJeztah commented:
@yongtang have you found anything, or need help?

thaJeztah commented:
docs failure is not related;

```
04:40:16 ERROR: (503) 1 links to (http://support.oracle.com)
04:40:16 			 link http://support.oracle.com on page engine/installation/linux/oracle.md
04:40:42 Skip
```

yongtang commented:
Thanks @thaJeztah. Previously the build error was caused by the import of container pkg (to get the `DefaultStopTimeout`) in `api/client/container/stop.go` and `api/client/container/restart.go`. That triggered the build of almost everything (libnetowrk, etc.) on Mac/darwin. I removed the import to get the build on darwin.

thaJeztah commented:
ping @yongtang can you rebase? Then we can get this moving again :+1:

yongtang commented:
Thanks @thaJeztah, rebased :smiley:.

keverw commented:
Setting a stop-timeout per container sounds like it'd be useful. I was looking at the Swarm Rolling Restart command and noticed no way to set the timeout. Having a way to set it per container for all commands instead of using the default sounds perfect.

vdemeester commented:
@yongtang rebase needed again :sweat:

yongtang commented:
Thanks @vdemeester rebased. 😃

vdemeester commented:
LGTM 🐸
@yongtang needs another rebase :sob:

yongtang commented:
Thanks @vdemeester. The pull request has been rebased.

cpuguy83 commented on cmd/dockerd/daemon.go:
```diff
@@ -353,16 +353,22 @@ func (cli *DaemonCli) stop() {
 		d.Shutdown()
 		close(ch)
 	}()
+	if shutdownTimeout < 0 {
```
why would this be `< 0`?

yongtang replied:
@cpuguy83 If it is `< 0` then we will wait until daemon is shutdown (no timeout).

vdemeester commented:
@yongtang needs another rebase :sweat:

yongtang commented:
Thanks @vdemeester, rebased.

mlaventure commented:
LGTM, but this PR needs yet another rebase for the docs :sweat_smile:

yongtang commented:
Thanks @mlaventure for the review. The PR has been rebased and updated. Please let me know if there are any other issues 😅 .

thaJeztah reviewed (approved):
docs LGTM, thanks!

ping @albers @sdurrheimer for completion scripts :heart:

vdemeester reviewed (approved):
LGTM 🐸

thaJeztah commented:
:tada: yeah! Sorry it took so long @yongtang - thanks so much!

yongtang commented:
Thanks all for the review and help!

arvenil commented:
Just to be double sure, can I specify `docker run --stop-timeout=-1` to force docker to wait infinitely ? And in such case, if that "infinity" never happens, then what option I have to stop container? Some "force" option? Or I can overwrite it with `docker stop --time=0`?
